File: react-native/react-native-pulsar/deps/Pulsar/Sources/Pulsar/Presets/SystemPresetsImpl.swift
import UIKit
import Foundation

@objc public class SystemImpactLightPreset : Player, Preset {
  public static let name: String = "SystemImpactLight"
  private var impactFeedbackGenerator: UIImpactFeedbackGenerator!

  @objc public init(_ haptics: Pulsar) {
//CODEGEN_BEGIN_{system_preset}
    super.init(haptics, audioOnly: true, rawDiscretePattern: [
      [0, 0.55, 0.4]
    ])
//CODEGEN_END_{system_preset}
    self.impactFeedbackGenerator = UIImpactFeedbackGenerator(style: .light)
    self.impactFeedbackGenerator.prepare()
  }

  @objc public override func play() {
    guard isEnabled else { return }
    super.play()
    Task { [impactFeedbackGenerator = self.impactFeedbackGenerator] in
      await impactFeedbackGenerator.impactOccurred()
    }
  }


  public static func getInstance(haptics: Pulsar) -> Preset {
    return SystemImpactLightPreset(haptics)
  }
}

@objc public class SystemImpactMediumPreset : Player, Preset {
  public static let name: String = "SystemImpactMedium"
  private var impactFeedbackGenerator: UIImpactFeedbackGenerator!

  public init(_ haptics: Pulsar) {
//CODEGEN_BEGIN_{system_preset}
    super.init(haptics, audioOnly: true, rawDiscretePattern: [
      [0, 0.7, 0.3]
    ])
//CODEGEN_END_{system_preset}
    self.impactFeedbackGenerator = UIImpactFeedbackGenerator(style: .medium)
    self.impactFeedbackGenerator.prepare()
  }

  @objc public override func play() {
    guard isEnabled else { return }
    super.play()
    Task { [impactFeedbackGenerator = self.impactFeedbackGenerator] in
      await impactFeedbackGenerator.impactOccurred()
    }
  }


  public static func getInstance(haptics: Pulsar) -> Preset {
    return SystemImpactMediumPreset(haptics)
  }
}

@objc public class SystemImpactHeavyPreset : Player, Preset {
  public static let name: String = "SystemImpactHeavy"
  private var impactFeedbackGenerator: UIImpactFeedbackGenerator!

  public init(_ haptics: Pulsar) {
//CODEGEN_BEGIN_{system_preset}
    super.init(haptics, audioOnly: true, rawDiscretePattern: [
      [0, 1, 0.45]
    ])
//CODEGEN_END_{system_preset}
    self.impactFeedbackGenerator = UIImpactFeedbackGenerator(style: .heavy)
    self.impactFeedbackGenerator.prepare()
  }

  @objc public override func play() {
    guard isEnabled else { return }
    super.play()
    Task { [impactFeedbackGenerator = self.impactFeedbackGenerator] in
      await impactFeedbackGenerator.impactOccurred()
    }
  }


  public static func getInstance(haptics: Pulsar) -> Preset {
    return SystemImpactHeavyPreset(haptics)
  }
}

@objc public class SystemImpactSoftPreset : Player, Preset {
  public static let name: String = "SystemImpactSoft"
  private var impactFeedbackGenerator: UIImpactFeedbackGenerator!

  public init(_ haptics: Pulsar) {
//CODEGEN_BEGIN_{system_preset}
    super.init(haptics, audioOnly: true, rawDiscretePattern: [
      [0, 0.6, 0.1]
    ])
//CODEGEN_END_{system_preset}
    self.impactFeedbackGenerator = UIImpactFeedbackGenerator(style: .soft)
    self.impactFeedbackGenerator.prepare()
  }

  @objc public override func play() {
    guard isEnabled else { return }
    super.play()
    Task { [impactFeedbackGenerator = self.impactFeedbackGenerator] in
      await impactFeedbackGenerator.impactOccurred()
    }
  }


  public static func getInstance(haptics: Pulsar) -> Preset {
    return SystemImpactSoftPreset(haptics)
  }
}

@objc public class SystemImpactRigidPreset : Player, Preset {
  public static let name: String = "SystemImpactRigid"
  private var impactFeedbackGenerator: UIImpactFeedbackGenerator!

  public init(_ haptics: Pulsar) {
//CODEGEN_BEGIN_{system_preset}
    super.init(haptics, audioOnly: true, rawDiscretePattern: [
      [0, 0.8, 0.95]
    ])
//CODEGEN_END_{system_preset}
    self.impactFeedbackGenerator = UIImpactFeedbackGenerator(style: .rigid)
    self.impactFeedbackGenerator.prepare()
  }

  @objc public override func play() {
    guard isEnabled else { return }
    super.play()
    Task { [impactFeedbackGenerator = self.impactFeedbackGenerator] in
      await impactFeedbackGenerator.impactOccurred()
    }
  }


  public static func getInstance(haptics: Pulsar) -> Preset {
    return SystemImpactRigidPreset(haptics)
  }
}

@objc public class SystemNotificationSuccessPreset : Player, Preset {
  public static let name: String = "SystemNotificationSuccess"
  private var feedbackGenerator: UINotificationFeedbackGenerator!

  public init(_ haptics: Pulsar) {
//CODEGEN_BEGIN_{system_preset}
    super.init(haptics, audioOnly: true, rawDiscretePattern: [
      [0, 0.6, 0.6],
      [150, 1, 1]
    ])
//CODEGEN_END_{system_preset}
    self.feedbackGenerator = UINotificationFeedbackGenerator()
    self.feedbackGenerator.prepare()
  }

  @objc public override func play() {
    guard isEnabled else { return }
    super.play()
    Task { [feedbackGenerator = self.feedbackGenerator] in
      await feedbackGenerator.notificationOccurred(.success)
    }
  }


  public static func getInstance(haptics: Pulsar) -> Preset {
    return SystemNotificationSuccessPreset(haptics)
  }
}

@objc public class SystemNotificationWarningPreset : Player, Preset {
  public static let name: String = "SystemNotificationWarning"
  private var feedbackGenerator: UINotificationFeedbackGenerator!

  public init(_ haptics: Pulsar) {
  //CODEGEN_BEGIN_{system_preset}
    super.init(haptics, audioOnly: true, rawDiscretePattern: [
      [0, 0.95, 1],
      [150, 0.6, 0.9]
    ])
  //CODEGEN_END_{system_preset}
    self.feedbackGenerator = UINotificationFeedbackGenerator()
    self.feedbackGenerator.prepare()
  }

  @objc public override func play() {
    guard isEnabled else { return }
    super.play()
    Task { [feedbackGenerator = self.feedbackGenerator] in
      await feedbackGenerator.notificationOccurred(.warning)
    }
  }


  public static func getInstance(haptics: Pulsar) -> Preset {
    return SystemNotificationWarningPreset(haptics)
  }
}

@objc public class SystemNotificationErrorPreset : Player, Preset {
  public static let name: String = "SystemNotificationError"
  private var feedbackGenerator: UINotificationFeedbackGenerator!

  public init(_ haptics: Pulsar) {
//CODEGEN_BEGIN_{system_preset}
    super.init(haptics, audioOnly: true, rawDiscretePattern: [
      [0, 0.7, 0.5],
      [100, 0.7, 0.5],
      [200, 0.7, 0.8],
      [250, 0.8, 0.4]
    ])
//CODEGEN_END_{system_preset}
    self.feedbackGenerator = UINotificationFeedbackGenerator()
    self.feedbackGenerator.prepare()
  }

  @objc public override func play() {
    guard isEnabled else { return }
    super.play()
    Task { [feedbackGenerator = self.feedbackGenerator] in
      await feedbackGenerator.notificationOccurred(.error)
    }
  }


  public static func getInstance(haptics: Pulsar) -> Preset {
    return SystemNotificationErrorPreset(haptics)
  }
}

@objc public class SystemSelectionPreset : Player, Preset {
  public static let name: String = "SystemSelection"
  private var feedbackGenerator: UISelectionFeedbackGenerator!

  public init(_ haptics: Pulsar) {
//CODEGEN_BEGIN_{system_preset}
    super.init(haptics, audioOnly: true, rawDiscretePattern: [
      [0, 0.4, 0.7]
    ])
//CODEGEN_END_{system_preset}
    self.feedbackGenerator = UISelectionFeedbackGenerator()
    self.feedbackGenerator.prepare()
  }

  @objc public override func play() {
    guard isEnabled else { return }
    super.play()
    Task { [feedbackGenerator = self.feedbackGenerator] in
      await feedbackGenerator.selectionChanged()
    }
  }


  public static func getInstance(haptics: Pulsar) -> Preset {
    return SystemSelectionPreset(haptics)
  }
}

File: kmp/Pulsar/library/src/iosMain/kotlin/com/swmansion/pulsar/Presets/SystemPresetsImpl.kt
package com.swmansion.pulsar

import platform.UIKit.UIImpactFeedbackGenerator
import platform.UIKit.UIImpactFeedbackStyle
import platform.UIKit.UINotificationFeedbackGenerator
import platform.UIKit.UINotificationFeedbackType
import platform.UIKit.UISelectionFeedbackGenerator

internal class IOSSystemImpactLightPreset(haptics: IOSPulsarHandle) : IOSSystemImpactPreset(
    haptics = haptics,
    presetName = "SystemImpactLight",
    style = UIImpactFeedbackStyle.UIImpactFeedbackStyleLight,
    rawDiscretePattern = listOf(listOf(0f, 0.55f, 0.4f)),
)

internal class IOSSystemImpactMediumPreset(haptics: IOSPulsarHandle) : IOSSystemImpactPreset(
    haptics = haptics,
    presetName = "SystemImpactMedium",
    style = UIImpactFeedbackStyle.UIImpactFeedbackStyleMedium,
    rawDiscretePattern = listOf(listOf(0f, 0.7f, 0.3f)),
)

internal class IOSSystemImpactHeavyPreset(haptics: IOSPulsarHandle) : IOSSystemImpactPreset(
    haptics = haptics,
    presetName = "SystemImpactHeavy",
    style = UIImpactFeedbackStyle.UIImpactFeedbackStyleHeavy,
    rawDiscretePattern = listOf(listOf(0f, 1f, 0.45f)),
)

internal class IOSSystemImpactSoftPreset(haptics: IOSPulsarHandle) : IOSSystemImpactPreset(
    haptics = haptics,
    presetName = "SystemImpactSoft",
    style = UIImpactFeedbackStyle.UIImpactFeedbackStyleSoft,
    rawDiscretePattern = listOf(listOf(0f, 0.6f, 0.1f)),
)

internal class IOSSystemImpactRigidPreset(haptics: IOSPulsarHandle) : IOSSystemImpactPreset(
    haptics = haptics,
    presetName = "SystemImpactRigid",
    style = UIImpactFeedbackStyle.UIImpactFeedbackStyleRigid,
    rawDiscretePattern = listOf(listOf(0f, 0.8f, 0.95f)),
)

internal class IOSSystemNotificationSuccessPreset(haptics: IOSPulsarHandle) : IOSSystemNotificationPreset(
    haptics = haptics,
    presetName = "SystemNotificationSuccess",
    type = UINotificationFeedbackType.UINotificationFeedbackTypeSuccess,
    rawDiscretePattern = listOf(
        listOf(0f, 0.6f, 0.6f),
        listOf(150f, 1f, 1f),
    ),
)

internal class IOSSystemNotificationWarningPreset(haptics: IOSPulsarHandle) : IOSSystemNotificationPreset(
    haptics = haptics,
    presetName = "SystemNotificationWarning",
    type = UINotificationFeedbackType.UINotificationFeedbackTypeWarning,
    rawDiscretePattern = listOf(
        listOf(0f, 0.95f, 1f),
        listOf(150f, 0.6f, 0.9f),
    ),
)

internal class IOSSystemNotificationErrorPreset(haptics: IOSPulsarHandle) : IOSSystemNotificationPreset(
    haptics = haptics,
    presetName = "SystemNotificationError",
    type = UINotificationFeedbackType.UINotificationFeedbackTypeError,
    rawDiscretePattern = listOf(
        listOf(0f, 0.7f, 0.5f),
        listOf(100f, 0.7f, 0.5f),
        listOf(200f, 0.7f, 0.8f),
        listOf(250f, 0.8f, 0.4f),
    ),
)

internal class IOSSystemSelectionPreset(haptics: IOSPulsarHandle) : IOSPlayer(
    haptics = haptics,
    audioOnly = true,
    rawDiscretePattern = listOf(listOf(0f, 0.4f, 0.7f)),
) {
    private val feedbackGenerator = UISelectionFeedbackGenerator()

    override val name: String = "SystemSelection"

    init {
        feedbackGenerator.prepare()
    }

    override fun play() {
        if (!isEnabled) return
        super.play()
        feedbackGenerator.selectionChanged()
    }
}

internal open class IOSSystemImpactPreset(
    haptics: IOSPulsarHandle,
    private val presetName: String,
    style: UIImpactFeedbackStyle,
    rawDiscretePattern: List<List<Float>>,
) : IOSPlayer(
    haptics = haptics,
    audioOnly = true,
    rawDiscretePattern = rawDiscretePattern,
) {
    private val impactFeedbackGenerator = UIImpactFeedbackGenerator(style = style)

    override val name: String = presetName

    init {
        impactFeedbackGenerator.prepare()
    }

    override fun play() {
        if (!isEnabled) return
        super.play()
        impactFeedbackGenerator.impactOccurred()
    }
}

internal open class IOSSystemNotificationPreset(
    haptics: IOSPulsarHandle,
    private val presetName: String,
    private val type: UINotificationFeedbackType,
    rawDiscretePattern: List<List<Float>>,
) : IOSPlayer(
    haptics = haptics,
    audioOnly = true,
    rawDiscretePattern = rawDiscretePattern,
) {
    private val feedbackGenerator = UINotificationFeedbackGenerator()

    override val name: String = presetName

    init {
        feedbackGenerator.prepare()
    }

    override fun play() {
        if (!isEnabled) return
        super.play()
        feedbackGenerator.notificationOccurred(type)
    }
}
